File: dynamic_listings/python/default/collisiondetection/pixel_perfect.py
class Color:

    """
    Defines a color
    """

    def __init__(self, colorData: int = 0x0) -> None:
        """Initializes the color """
        self.colorData: int = colorData

    def is_white(self) -> bool:
        return self.colorData == 0xFFFFFF


class Bitmask:

    """A sprite bitmask"""

    def __init__(self, data: list[Color] = None) -> None:
        """Initializes the bitmask with an array of colors

        :data: An array of colors

        """
        self._data: list[Color] = data

    def get_color(self, x: int, y: int) -> Color:
        """
        Gets the color at a certain x and y of the bitmap

        :x: x coordinate of the pixel to return
        :y: y coordinate of the pixel to return
        :returns: The color found

        """
        # Just an example
        ...

class Sprite:
    def __init__(self, bitmask: Bitmask, x: int, y: int, width: int, height: int) -> None:
        self.bitmask: Bitmask = bitmask
        self.x: int = x
        self.y: int = y
        self.width: int = width
        self.height: int = height

def pixel_perfect_collision(A: Sprite, B: Sprite) -> bool:
    """
    Calculates a pixel-perfect collsion

    :A: First sprite
    :B: Second Sprite
    :returns: A boolean telling us if the two sprites collide

    """
    # Calculate the intersecting rectangle to limit checks
    x1: int = max(A.x, B.x)
    x2: int = min((A.x + A.width), (B.x + B.width))

    y1: int = max(A.y, B.y)
    y2: int = min((A.y + A.height), (B.y + B.height))

    # For each pixel in the intersecting rectangle, let's check
    for y in range(y1, y2):
        for x in range(x1, x2):
            a: Color = A.bitmask.get_color(x - A.x, y - A.y)
            b: Color = B.bitmask.get_color(x - B.x, y - B.y)

            if (a.isWhite() and b.isWhite()):
                return True

    # If no collision has occurred by the end of the checking, we're safe
    return False

File: dynamic_listings/rust/default/collisiondetection/pixel_perfect.rs
struct Color {
    color_data: u32,
}

impl Color {
    fn is_white(&self) -> bool {
        //...
    }
}

struct Bitmask {
    data: Vec<Color>,
}

impl Bitmask {
    fn color(&self, x: u32, y: u32) -> Color {
        // ...
    }
}

struct Sprite {
    bitmask: Bitmask,
    x: u32,
    y: u32,
    width: u32,
    height: u32,
}

fn pixel_perfect_collision(a: &Sprite, b: &Sprite) -> bool {
    // Calculate the intersecting rectangle to limit checks
    let x1 = u32::max(a.x, b.x);
    let x2 = u32::min(a.x + a.width, b.x + b.width);

    let y1 = u32::max(a.y, b.y);
    let y2 = u32::min(a.y + a.height, b.y + b.height);

    // For each pixel in the intersecting rectangle, let's check
    for y in y1..y2 {
        for x in x1..x2 {
            // We're working in the intersecting triangle, so we'll need to
            // rework our coordinates
            let a = a.bitmask.color(x - a.x, y - a.y);
            let b = b.bitmask.color(x - b.x, y - b.y);
            if a.is_white() && b.is_white() {
                return true;
            }
        }
    }

    // If no collision is occurred by the end of the checking, we're safe
    false
}
